File: Services/StartupService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;
using Microsoft.Win32;

namespace InvolveX.Cli.Services
{
    public class StartupService
    {
        private readonly LogService _logService;

        public StartupService(LogService logService)
        {
            _logService = logService;
        }

        public async Task<List<string>> ListStartupPrograms()
        {
            _logService.Log("Listing startup programs from registry and Task Scheduler.");
            var startupPrograms = new List<string>();

            try
            {
                // Check Run registry keys
                var runKeys = new[]
                {
                    @"SOFTWARE\Microsoft\Windows\CurrentVersion\Run",
                    @"SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce",
                    @"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Run",
                    @"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\RunOnce"
                };

                foreach (var keyPath in runKeys)
                {
                    using (var key = Registry.LocalMachine.OpenSubKey(keyPath))
                    {
                        if (key != null)
                        {
                            foreach (var valueName in key.GetValueNames())
                            {
                                var value = key.GetValue(valueName);
                                if (value != null)
                                {
                                    startupPrograms.Add($"Registry ({keyPath}): {valueName} -> {value}");
                                }
                            }
                        }
                    }

                    using (var key = Registry.CurrentUser.OpenSubKey(keyPath))
                    {
                        if (key != null)
                        {
                            foreach (var valueName in key.GetValueNames())
                            {
                                var value = key.GetValue(valueName);
                                if (value != null)
                                {
                                    startupPrograms.Add($"Registry (CurrentUser): {valueName} -> {value}");
                                }
                            }
                        }
                    }
                }

                // Check Startup folder
                var startupFolder = Environment.GetFolderPath(Environment.SpecialFolder.Startup);
                if (System.IO.Directory.Exists(startupFolder))
                {
                    var files = System.IO.Directory.GetFiles(startupFolder);
                    foreach (var file in files)
                    {
                        startupPrograms.Add($"Startup Folder: {System.IO.Path.GetFileName(file)}");
                    }
                }

                // Check Task Scheduler startup tasks
                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "schtasks",
                        Arguments = "/query /tn \"\\Microsoft\\Windows\\*\" /fo csv /nh",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                process.Start();
                string output = await process.StandardOutput.ReadToEndAsync();
                string error = await process.StandardError.ReadToEndAsync();
                await process.WaitForExitAsync();

                if (process.ExitCode == 0)
                {
                    var lines = output.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
                    foreach (var line in lines)
                    {
                        if (!string.IsNullOrWhiteSpace(line) && line.Contains("Ready") || line.Contains("Running"))
                        {
                            var parts = line.Split(',');
                            if (parts.Length > 0)
                            {
                                startupPrograms.Add($"Task Scheduler: {parts[0].Trim('\"')}");
                            }
                        }
                    }
                }

                _logService.Log($"Found {startupPrograms.Count} startup programs.");
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception listing startup programs: {ex.Message}");
                startupPrograms.Add($"Error: {ex.Message}");
            }

            return startupPrograms;
        }

        public async Task<bool> DisableStartupProgram(string programName)
        {
            _logService.Log($"Attempting to disable startup program: {programName}");

            try
            {
                // This is a simplified implementation. In practice, you'd need to identify
                // the source (registry, startup folder, task scheduler) and handle accordingly
                var runKeys = new[]
                {
                    @"SOFTWARE\Microsoft\Windows\CurrentVersion\Run",
                    @"SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce",
                    @"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Run",
                    @"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\RunOnce"
                };

                foreach (var keyPath in runKeys)
                {
                    using (var key = Registry.LocalMachine.OpenSubKey(keyPath, true))
                    {
                        if (key != null && Array.Exists(key.GetValueNames(), name => name.Equals(programName, StringComparison.OrdinalIgnoreCase)))
                        {
                            key.DeleteValue(programName);
                            _logService.Log($"Disabled startup program from registry: {programName}");
                            return true;
                        }
                    }

                    using (var key = Registry.CurrentUser.OpenSubKey(keyPath, true))
                    {
                        if (key != null && Array.Exists(key.GetValueNames(), name => name.Equals(programName, StringComparison.OrdinalIgnoreCase)))
                        {
                            key.DeleteValue(programName);
                            _logService.Log($"Disabled startup program from registry: {programName}");
                            return true;
                        }
                    }
                }

                // For Task Scheduler items, this would require more complex logic
                // For now, we'll log that manual intervention may be needed
                _logService.Log($"Could not automatically disable: {programName}. Manual intervention may be required.");
                return false;
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception disabling startup program {programName}: {ex.Message}");
                return false;
            }
        }
    }
}

File: services/StartupService.js
const { spawn } = require('child_process');

class StartupService {
  constructor(logService) {
    this.logService = logService;
  }

  async listStartupPrograms() {
    this.logService.log('Listing startup programs from registry and Task Scheduler.');
    const startupPrograms = [];

    try {
      // Check registry Run keys using PowerShell
      const psScript = `
                # Get registry startup programs
                $runKeys = @(
                    "HKLM:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKLM:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKLM:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKLM:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKCU:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKCU:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKCU:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKCU:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\RunOnce"
                )

                foreach ($keyPath in $runKeys) {
                    try {
                        if (Test-Path $keyPath) {
                            $properties = Get-ItemProperty -Path $keyPath -ErrorAction SilentlyContinue
                            if ($properties) {
                                $properties.PSObject.Properties | Where-Object { $_.Name -notlike 'PS*' } | ForEach-Object {
                                    Write-Output "Registry ($keyPath): $($_.Name) -> $($_.Value)"
                                }
                            }
                        }
                    } catch {
                        # Skip inaccessible keys
                    }
                }

                # Get startup folder contents
                $startupFolder = [Environment]::GetFolderPath('Startup')
                if (Test-Path $startupFolder) {
                    Get-ChildItem -Path $startupFolder | ForEach-Object {
                        Write-Output "Startup Folder: $($_.Name)"
                    }
                }

                # Get Task Scheduler startup tasks
                try {
                    $tasks = Get-ScheduledTask | Where-Object { $_.TaskPath -like '\\Microsoft\\Windows\\*' -and $_.State -ne 'Disabled' }
                    foreach ($task in $tasks) {
                        Write-Output "Task Scheduler: $($task.TaskName)"
                    }
                } catch {
                    Write-Output "Task Scheduler: Could not retrieve tasks"
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);

      if (result.code === 0) {
        const lines = result.stdout.split('\n').filter(line => line.trim());
        startupPrograms.push(...lines);
      }

      this.logService.log(`Found ${startupPrograms.length} startup programs.`);
    } catch (error) {
      this.logService.log(`Exception listing startup programs: ${error.message}`);
      startupPrograms.push(`Error: ${error.message}`);
    }

    return startupPrograms;
  }

  async disableStartupProgram(programName) {
    this.logService.log(`Attempting to disable startup program: ${programName}`);

    try {
      // Try to disable from registry first
      const psScript = `
                $programName = "${programName}"
                $found = $false

                $runKeys = @(
                    "HKLM:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKLM:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKLM:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKLM:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKCU:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKCU:\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\RunOnce",
                    "HKCU:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Run",
                    "HKCU:\\SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\RunOnce"
                )

                foreach ($keyPath in $runKeys) {
                    try {
                        if (Test-Path $keyPath) {
                            $value = Get-ItemProperty -Path $keyPath -Name $programName -ErrorAction SilentlyContinue
                            if ($value) {
                                Remove-ItemProperty -Path $keyPath -Name $programName -ErrorAction SilentlyContinue
                                Write-Output "SUCCESS: Disabled from registry"
                                $found = $true
                                break
                            }
                        }
                    } catch {
                        # Continue to next key
                    }
                }

                if (-not $found) {
                    # Try to disable scheduled task
                    try {
                        $task = Get-ScheduledTask -TaskName $programName -ErrorAction SilentlyContinue
                        if ($task) {
                            Disable-ScheduledTask -TaskName $programName -ErrorAction SilentlyContinue
                            Write-Output "SUCCESS: Disabled scheduled task"
                            $found = $true
                        }
                    } catch {
                        # Task not found or couldn't be disabled
                    }
                }

                if (-not $found) {
                    Write-Output "NOT_FOUND"
                }
            `;

      const result = await this.runProcess('powershell', ['-Command', psScript]);

      if (result.code === 0 && result.stdout.includes('SUCCESS')) {
        this.logService.log(`Successfully disabled startup program: ${programName}`);
        return true;
      } else {
        this.logService.log(
          `Could not automatically disable: ${programName}. Manual intervention may be required.`
        );
        return false;
      }
    } catch (error) {
      this.logService.log(`Exception disabling startup program ${programName}: ${error.message}`);
      return false;
    }
  }

  async runProcess(command, args) {
    return new Promise((resolve, reject) => {
      const process = spawn(command, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true,
      });

      let stdout = '';
      let stderr = '';

      process.stdout.on('data', data => {
        stdout += data.toString();
      });

      process.stderr.on('data', data => {
        stderr += data.toString();
      });

      process.on('close', code => {
        resolve({
          code,
          stdout,
          stderr,
        });
      });

      process.on('error', error => {
        reject(error);
      });
    });
  }
}

module.exports = StartupService;
